### ISSUE TITLE
stream client endpoint plugins should be passed a reactor, but aren't

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5069]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5069|
|Type|defect|
|Created|2011-04-26 19:04:12Z|
|Branch|https://github.com/twisted/twisted/tree/parseStreamClient-reactor-5069-4|

As originally described on [#4956](https://github.com/twisted/twisted/issues/4956), stream server string endpoint plugins get a reactor passed to them correctly, but stream client string endpoint plugins don't.

We should fix that with a new interface and a deprecation and soforth.

Attachments:

* [tcp6-client-plugin.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/963/9639c68eef083d09baa29bd9117717393c61efb3/4de7b348c856164fe9ae7c712f550447dfd9b13f.diff) (3463 bytes) - added by ashfall on 2013-07-31 11:07:21Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5069 5069
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_parseStreamClient_reactor_5069_4 branches/parseStreamClient-reactor-5069-4
branch_author__habnabit habnabit
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1303844652000000 1303844652000000
changetime__1392920916000000 1392920916000000
version__None None
owner__habnabit habnabit
cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ccx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ccx commented|
|-|-|

The issue is there actually _two_ interfaces, one for the actual ISomethingEndpoint and one for the somethingFromString() methods. The latter is currently clumped into the first if and only if the implementation is a plugin. Current implementation of builtin endpoints does not conform to the interface used by *FromString, instead it is implemented as a special case with it's own lookup tables.

There are two ways to remedy this: 1) the "pure" way of separating the interfaces:

```python
class ISomethingEndpointPlugin(IPlugin):

    prefix = Attribute("the part before the colon it string specification")

    def createEndpoint(reactor, *args, **kwargs):
        """
        Create the actual endpoint from the information in the string passed as
        args and kwargs.
        """
```

1b) same as above, just make another attribute instead of having couple of "Something" variants, so the interface is generic to any *FromString function.

2) Make the attribute and the form of constructor mandatory for all implementation of endpoints. The question is: do we want to allow such implementation of endpoint that would be inherently incompatible with invocation from *FromString? For example Endpoint that would represent new connection over established ssh channel, as that needs a reference to the object representing the ssh connection, which is not reasonably string-encodable. On the other hand, just having the ```__init__``` always in the form of ```(reactor, *args, **kwargs)``` does not hurt anything, we just get useless attribute declared this way.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I'm somewhat confused by that comment.  For example, for stream client endpoints, there are already two interfaces: `IStreamClientEndpoint` and `IStreamClientEndpointStringParser`.

The problem in the ticket description seems pretty straightforward: the stream client endpoint parser doesn't get the reactor parameter passed to `clientFromString`.  There should be a way for it to get it.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_3"></a>|@habnabit commented|
|-|-|

(In [33794]) Branching to 'parseStreamClient-reactor-5069'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

(In [33812]) Deprecating the absence of a `reactor` argument.

Full test coverage over the new deprecation; tests pass. Also fixed a typo in
some adjacent clientFromString test docstrings.

Refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_5"></a>|@habnabit commented|
|-|-|

(In [33819]) Whoops, verbs!

Refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_6"></a>|@habnabit commented|
|-|-|

(In [33820]) Adding documentation about the new behavior and deprecation.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit commented|
|-|-|

Buildbot is all green except for the known intermittent failures.
[http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/parseStreamClient-reactor-5069]

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid set owner to @habnabit|
|-|-|

1.Should use ```deprecate.getDeprecationWarningString``` to construct the second argument to ```deprecate.warnAboutFunction``` see [twisted/internet/tcp.py#L202](twisted/internet/tcp.py#L202) for an example.
 1. You change ```IStreamClientEndpointStringParser``` which means that things already claiming to implement that interface will be lying when this change is done.  Obviously the existing interface is stupid and broken, but it's public so you can't really change it.
 1. Test coverage as the fix is implemented looks good, also there were no pyflakes problems or trailing whitespace and all of the lines were the right size.

Thanks for working diligently to improve the state of endpoints!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_10"></a>|@dreid commented|
|-|-|

Also there should be a News fragment per http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_11"></a>|@habnabit commented|
|-|-|

(In [33833]) Refactoring code out into t.i.endpoints._executeParseStreamClient.

Tests still pass; test coverage is still full.

Refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_12"></a>|@wallrj commented|
|-|-|

See duplicate ticket [#5870](https://github.com/twisted/twisted/issues/5870)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_13"></a>|@habnabit commented|
|-|-|

Going to revisit this when I get a chance. The approach I used before was cute, but it's much less terrible to make a new interface instead of trying to reuse the old one.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_14"></a>|@habnabit commented|
|-|-|

(In [39247]) Branching to 'parseStreamClient-reactor-5069-2'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_15"></a>|@habnabit commented|
|-|-|

(In [39250]) Add a new interface for string client parsers.

Refs [#5069](https://github.com/twisted/twisted/issues/5069).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_16"></a>|@habnabit commented|
|-|-|

(In [39252]) Deprecate the old interface.

Refs [#5069](https://github.com/twisted/twisted/issues/5069).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_17"></a>|@habnabit commented|
|-|-|

(In [39260]) Fix another typo.

Dang it's a long name to remember how to spell.

Refs [#5069](https://github.com/twisted/twisted/issues/5069).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_18"></a>|@habnabit removed owner|
|-|-|

Almost all green! A builder seems to have run out of disk space, pyflakes doesn't like the python 3 compatibility check in endpoints.py, and a fake plugin doesn't have a docstring (following the lead of other fakes in the same module. I could add it still, though).

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/parseStreamClient-reactor-5069-2

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)|@ashfall set owner to @ashfall<br>@ashfall set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_20"></a>|@ashfall removed owner<br>@ashfall set status to `new`|
|-|-|

Thanks for working on this! Also, this is by no means a complete review, hence I'm not removing the 'review' keyword.

I wrote a plugin for `TCP6ClientEndpoint` to see how things go.

```
@implementer(IPlugin, IStreamClientEndpointStringParserWithReactor)
class _TCP6ClientParser(object):
    prefix = "tcp6"
    def _parseClient(self, reactor, host, port):
        port = int(port)
        return TCP6ClientEndpoint(reactor, host, port)


    def parseStreamClient(self, reactor, *args, **kwargs):
        print "args = ", args          # Ugly. For debugging.
        print "reactor = ", reactor    # Again, for debugging.
        return self._parseClient(reactor, *args, **kwargs)
```

And the corresponding tests:
```
class TCP6ClientEndpointPluginTests(unittest.TestCase):
    """
    Unit tests for the TCP IPv6 stream client endpoint string description
    parser.
    """
    _parserClass = endpoints._TCP6ClientParser

    def test_pluginDiscovery(self):
        """
        L{endpoints._TCP6ClientParser} is found as a plugin for
        L{interfaces.IStreamClientEndpointStringParser} interface.
        """
        parsers = list(getPlugins(
            interfaces.IStreamClientEndpointStringParser))
        for p in parsers:
            if isinstance(p, self._parserClass):
                break
        else:
            self.fail(
                "Did not find TCP6ClientEndpoint parser in %r" % (parsers,))


    def test_interface(self):
        """
        L{endpoints._TCP6ClientParser} instances provide
        L{interfaces.IStreamClientEndpointStringParser}.
        """
        parser = self._parserClass()
        self.assertTrue(verifyObject(
            interfaces.IStreamClientEndpointStringParserWithReactor, parser))


    def test_stringDescription(self):
        """
        L{clientFromString} returns a L{TCP6ClientEndpoint} instance with a
        'tcp6' endpoint string description.
        """
        ep = endpoints.clientFromString(
                MemoryReactor(), "tcp6:\:\:1:8080") # detects the args properly if the string is "tcp6::\:\:1:8080"
        self.assertIsInstance(ep, endpoints.TCP6ClientEndpoint)
        self.assertIsInstance(ep._reactor, MemoryReactor)
        self.assertEqual(ep._host, "::1")
        self.assertEqual(ep._port, 8080)
        self.assertEqual(ep._timeout, 30)
        self.assertEqual(ep._bindAddress, None)
```

And trial gives the following output (only the results for the above test-case):
```
  TCP6ClientEndpointPluginTests
    test_interface ...                                                     [OK]
    test_pluginDiscovery ...                                               [OK]
    test_stringDescription ... args =  ('8080',)
reactor =  ::1
                                         [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/ashfall/Projects/Twisted/branches/parseStreamClient-reactor-5069-2/twisted/internet/test/test_endpoints.py", line 2619, in test_stringDescription
    MemoryReactor(), "tcp6:\:\:1:8080") # detects the args properly if the string is "tcp6::\:\:1:8080"
  File "/home/ashfall/Projects/Twisted/branches/parseStreamClient-reactor-5069-2/twisted/internet/endpoints.py", line 1612, in clientFromString
    return plugin.parseStreamClient(*args, **kwargs)
  File "/home/ashfall/Projects/Twisted/branches/parseStreamClient-reactor-5069-2/twisted/internet/endpoints.py", line 1110, in parseStreamClient
    return self._parseClient(reactor, *args, **kwargs)
exceptions.TypeError: _parseClient() takes exactly 4 arguments (3 given)

twisted.internet.test.test_endpoints.TCP6ClientEndpointPluginTests.test_stringDescription
-------------------------------------------------------------------------------
```

which indicates(I think) something in `clientFromString` isn't working as expected. On trying various variants, the plugin, as constructed above, ignores the first parameter in `clientFromString` (i.e. MemoryReactor() in `test_stringDescription` above). I suspect it's in some way related to `self.assertEqual(clientEndpoint.args, (reactor, 'alpha', 'beta'))` in `test_stringParserWithReactor`.

I may have overlooked something while writing the tests and/or the plugin itself, I'm afraid I am running short of time to verify (apologies).

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_21"></a>|@ashfall commented|
|-|-|

I took another look at this, and it looks like the error was because of dropin.cache. Deleting it and re-running trial fixed it, all tests pass. Attaching a diff for the same, just in case.

Replying to [ashfall](#note_20):


>     def test_pluginDiscovery(self):
>         """
>         L{endpoints._TCP6ClientParser} is found as a plugin for
>         L{interfaces.IStreamClientEndpointStringParser} interface.
>         """
>         parsers = list(getPlugins(
>             interfaces.IStreamClientEndpointStringParser))

This should instead be `interfaces.IStreamClientEndpointStringParserWithReactor`.


The plugin works as it should, the test coverage for the new interface looks good, I guess it's time for someone else to take a look! :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @habnabit<br>@exarkun set status to `new`|
|-|-|

Thanks!  It will be great to see this bug squashed and to be able to actually write testable endpoint parsers (not to mention parsers that do what `clientFromString` actually claims they'll do).

  1. I find the text of the warning is a little awkward.  I don't know if people will understand "used for" in this context.  The plugin's `parseStreamClient` is _implementing_ a method of the interface.  Maybe that verb could be worked into the text somehow?
  1. The `IStreamClientEndpointStringParser` attribute of `twisted.internet.interfaces` should be deprecated as well.  In fact, maybe it should be the only thing that's deprecated.  Kudos for figuring out `warnAboutFunction` in the plugin discovery code but it seems like emitting the warning at whatever code imports the interface will get the job done (and will result in warnings coming up sooner and more reliably, since it won't depend on someone actually traversing the plugins).
  1. The `IStreamClientEndpointStringParserWithReactor` docstring talks about "strings" and some of the attribute docs talk about "str".  Please make this unambiguous by using "byte strings" and "bytes" instead (don't forget to use b"" literals in the examples).
  1. Also use L{} instead of C{} to refer to builtin types.
  1. When using an interface as a type in epytext markup (eg for `rtype`) be sure to say "provider" or "implementer"
  1. New code should use `zope.interface.implementer` instead of `zope.interface.implements`.
  1. Where possible, it's nicer to only have one assertion in a test method.  Sometimes this becomes awkward or infeasible (as is the case in the tests for the deprecation warning, I think) but for cases like `test_stringParserWithReactorHasPreference` I think it's doable.  You can make all of the comparisons with a single `assertEqual` of a tuple of the individual values.
  1. All of the existing implementations of the now-deprecated interface that are part of Twisted itself should be switched to the new interface.  This should be done 'before' the deprecation is added - so part of this ticket, probably.

Thanks again!  Very sorry about the long review round-trip. :(

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_24"></a>|@habnabit commented|
|-|-|

(In [40181]) Branching to 'parseStreamClient-reactor-5069-3'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_25"></a>|@habnabit commented|
|-|-|

(In [40182]) Merge forward. refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_26"></a>|@habnabit set status to `assigned`|
|-|-|

Just a few remarks as I'm making changes.

 1. Fixed this by removing the deprecation and implementing point 2 instead.
 2. Switched to using deprecatedModuleAttribute for IStreamClientEndpointStringParser.
 3. As far as I can tell, the type expected by `endpoints._parse` is `str` in both 2.x and 3.x. At any rate, `endpoints._parse` fails when called with `bytes` on python 3. It sounds like this is unintentional, but should this current behavior be reflected in the docs?
 4. Fixed.
 5. Also fixed.
 6. Also fixed.
 7. Changed the tests to use `assertEqual` with a tuple.
 8. At the moment, there are no non-test implementations of `IStreamClientEndpointStringParser` in Twisted--that's probably how this went unnoticed in the first place.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_28"></a>|@habnabit commented|
|-|-|

(In [40628]) Address review comments. refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_29"></a>|@habnabit commented|
|-|-|

(In [40629]) Update docstrings to reflect endpoint descriptions are bytes.

refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_30"></a>|@habnabit commented|
|-|-|

(In [40630]) Bump 13.2 -> 14.0.

refs [#5069](https://github.com/twisted/twisted/issues/5069).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_31"></a>|@habnabit commented|
|-|-|

(In [40631]) Prevent some test pollution.

The dropin.cache being generated in some tests was being picked up in other,
unrelated tests. This wouldn't normally be a problem, but unpickling an
IStreamClientEndpointStringParser plugin triggers a warning, and the test in
particular ensured that there was only one warning triggered. Removing the
dropin.cache file prevents it from being loaded entirely.

For reference, the failing test was
twisted.trial.test.test_script.WithoutModuleTests.test_disableAlreadyImportedModule.

refs [#5069](https://github.com/twisted/twisted/issues/5069).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_32"></a>|@habnabit commented|
|-|-|

(In [40632]) Fix a bug in addFakePlugin.

Previously it tried to save and restore __path__, but didn't actually copy the
old __path__. As a result, the changes to __path__ never got reverted.

This also removes my old attempt at stopping test pollution in a different way.

refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_33"></a>|@habnabit removed owner<br>@habnabit set status to `new`|
|-|-|

[Buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/parseStreamClient-reactor-5069-3) looks mostly okay. There's a false positive on pyflakes and I'm not sure whether it's really necessary to provide docstrings for all of the fake parser plugins. (Or is it?) All of the other builders look good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun set owner to @habnabit|
|-|-|

Thanks!

  1. twisted/internet/endpoints.py
    1. The example that now reads `serverFromString(b"ssl:443:privateKey=%s" % (myPathName,))` is missing `reactor` as the first argument.
    1. Likewise for the example about 10 lines below it.
  1. twisted/internet/interfaces.py
    1. The return type for `IStreamClientEndpointStringParserWithReactor.parseStreamClient` should be documented as _an L{IStreamClientEndpoint} provider_.  If a class `implements(...)`  an interface then the **class** implements the interface. :)  Instances of the class _provide_ the interface.
  1. twisted/internet/test/test_endpoints.py
    1. Tests, eg `test_stringParserWithReactor`, should use byte string literals to pass in byte strings.
  1. "At the moment, there are no non-test implementations of IStreamClientEndpointStringParser in Twisted--that's probably how this went unnoticed in the first place." - if you want, maybe you could replace one of the hard-coded client parsers with a plugin?  Of course we can always do that later, but it might not hurt to do as part of this change to prove this really works. :)  Up to you, though.
  1. "As far as I can tell, the type expected by endpoints._parse is str in both 2.x and 3.x. At any rate, endpoints._parse fails when called with bytes on python 3. It sounds like this is unintentional, but should this current behavior be reflected in the docs?" - I think that's explained by the fact that this code hasn't been ported yet.  Part of the idea with using explicit byte string literals everywhere is that it should make that porting work easier when it happens (since it prevents the type of half the objects involved from randomly changing out from under you).

The rest looks great.  Please address those points, check buildbot again, and then merge.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_35"></a>|@habnabit commented|
|-|-|

(In [41732]) Address review comments. refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_36"></a>|@habnabit commented|
|-|-|

(In [41733]) Add more 'b' prefixes. refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_37"></a>|@habnabit commented|
|-|-|

(In [41734]) Branching to 'parseStreamClient-reactor-5069-4'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_38"></a>|@habnabit commented|
|-|-|

(In [41735]) Merge forward. refs [#5069](https://github.com/twisted/twisted/issues/5069)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_39"></a>|@habnabit commented|
|-|-|

[Buildbot looks good http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/parseStreamClient-reactor-5069-4]. The failures in [one Windows XP builder http://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/3339] appear to be intermittent and crop up in the build results for other branches as well. The only twistedchecker error comes from a fake plugin implementation which is only used for tests.

Merging!

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_40"></a>|@habnabit commented|
|-|-|

wow, I really need to remember to use trac's preview feature more often, because I still suck at trac markup.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_41"></a>|@habnabit set status to `closed`|
|-|-|

(In [41738]) Merge parseStreamClient-reactor-5069-4: Add IStreamClientEndpointStringParserWithReactor, deprecating IStreamClientEndpointStringParser.

Author: habnabit
Reviewers: exarkun, ashfall
Fixes: [#5069](https://github.com/twisted/twisted/issues/5069)
Refs: [#5720](https://github.com/twisted/twisted/issues/5720)

The new interface IStreamClientEndpointStringParserWithReactor will supply the
reactor to its parseStreamClient method, passed along from 
twisted.internet.endpoints.clientFromString. The old interface, 
IStreamClientEndpointStringParser, which did not supply the reactor, is now 
deprecated.
